File: packages/js/src/reporter.js
export class Reporter {
  constructor(format = 'text') {
    this.format = format;
  }

  printValidation(result) {
    if (this.format === 'json') {
      console.log(JSON.stringify(result, null, 2));
      return;
    }

    console.log(`${result.pass ? 'PASS' : 'FAIL'} ${result.source} (${result.duration}ms)`);
    console.log(`Layers: ${result.availableLayers.length}  Features: ${result.totalFeatures}`);
    for (const [name, layer] of Object.entries(result.layers)) {
      console.log(`  ${layer.valid ? 'ok' : 'xx'} ${name}: ${layer.featureCount}`);
    }
    for (const error of result.errors) console.error(`ERROR ${error.code}: ${error.message}`);
    for (const warning of result.warnings) console.warn(`WARN ${warning.code}: ${warning.message}`);
  }

  printValidationBatch(results) {
    if (this.format === 'json') {
      console.log(JSON.stringify(results, null, 2));
      return;
    }

    const passed = results.filter((result) => result.pass).length;
    console.log(`TileGuard batch: ${passed}/${results.length} passing`);
    for (const result of results) this.printValidation(result);
  }

  printLint(result) {
    if (this.format === 'json') {
      console.log(JSON.stringify(result, null, 2));
      return;
    }
    console.log(`${result.pass ? 'PASS' : 'FAIL'} ${result.source} (${result.duration}ms)`);
    for (const error of result.errors) console.error(`ERROR ${error.code}: ${error.message}`);
    for (const warning of result.warnings) console.warn(`WARN ${warning.code}: ${warning.message}`);
  }

  printRender(result) {
    if (this.format === 'json') {
      console.log(JSON.stringify(result, null, 2));
      return;
    }
    console.log(`${result.pass ? 'PASS' : 'FAIL'} ${result.fixture} (${result.duration}ms)`);
    if (result.message) console.log(result.message);
    for (const error of result.errors || []) console.error(`ERROR ${error.code}: ${error.message}`);
  }

  printRenderSummary(results) {
    if (this.format === 'json') {
      console.log(JSON.stringify(results, null, 2));
      return;
    }
    const passed = results.filter((result) => result.pass).length;
    console.log(`Render fixtures: ${passed}/${results.length} passing`);
    for (const result of results) this.printRender(result);
  }
}

File: packages/python/tileguard/reporter.py
from __future__ import annotations

import json

from .validate import ValidationResult


def print_validation(result: ValidationResult, fmt: str = "text") -> None:
    if fmt == "json":
        print(json.dumps(result.to_dict(), indent=2))
        return

    print(f"{'PASS' if result.pass_ else 'FAIL'} {result.source} ({result.duration}ms)")
    print(f"Layers: {len(result.available_layers)}  Features: {result.total_features}")
    for name, layer in result.layers.items():
        marker = "ok" if layer.valid else "xx"
        print(f"  {marker} {name}: {layer.feature_count}")
    for error in result.errors:
        print(f"ERROR {error['code']}: {error['message']}")
    for warning in result.warnings:
        print(f"WARN {warning['code']}: {warning['message']}")


def print_validation_batch(results: list[ValidationResult], fmt: str = "text") -> None:
    if fmt == "json":
        print(json.dumps([result.to_dict() for result in results], indent=2))
        return
    passed = sum(1 for result in results if result.pass_)
    print(f"TileGuard batch: {passed}/{len(results)} passing")
    for result in results:
        print_validation(result, fmt="text")
